extras/usb_id/usb_id.8
.TH USB_ID 8 "January 2009" "" "Linux Administrator's Manual"
.SH NAME
usb_id \- udev callout to read information from USB devices
.SH SYNOPSIS
.B usb_id
[\fB--usb-info\fP] [\fB--num-info\fP] [\fB--export\fP] [\fB--debug\fP]
.I devpath
.br
.B usb_id --help
.SH "DESCRIPTION"
.B usb_id
is normally called from a udev rule, to provide udev with a unique string and
additional information (uuid, label) for a USB device. Udev can use this
information to create symlinks in /dev to the real device node.
.SH USAGE
.B usb_id \fIdevpath\fP
opens the device node specified at the commandline and prints the
information chosen by the options.
.SH OPTIONS
The following commandline switches are supported to specify what usb_id
should print:
.TP
.B -u, --usb-info
print USB strings instead of, e.g. SCSI strings for a mass storage device
.TP
.B -n, --num-info
print numerical values instead of strings. Implies \fB--usb-info\fP.
.TP
.B -x, --export
print values as environment keys
.TP
.B -h, --help
print usage help
.RE
.SH SEE ALSO
.BR udev (7)
.SH AUTHORS
Developed by Hannes Reinecke <hare@suse.de>.


NEWS
udev 141
========
Bugfixes.

udev 140
========
Bugfixes.

"udevadm settle" now optionally accepts a range of events to wait for,
instead of waiting for "all" events.

udev 139
========
Bugfixes.

The installed watch for block device metadata changes is now removed
during event hadling, because some (broken) tools may be called from udev
rules and (wrongly) open the device with write access. After the finished
event handling the watch is restored.

udev 138
========
Bugfixes.

Device nodes can be watched for changes with inotify with OPTIONS="watch".
If closed after being opened for writing, a "change" uevent will occur.
/dev/disk/by-{label,uuid}/* symlinks will be automatically updated.

udev 137
========
Bugfixes.

The udevadm test command has no longer a --force option, nodes and symlinks
are always updated with a test run now.

The udevd daemon can be started with --resolve-names=never to avoid all user
and group lookups (e.g. in cut-down systems) or --resolve-names=late to
lookup user and groups every time events are handled.

udev 136
========
Bugfixes.

We are currently merging the Ubuntu rules in the udev default rules,
and get one step closer to provide a common Linux /dev setup, regarding
device names, symlinks, and default device permissions. On udev startup,
we now expect the following groups to be resolvable to their ids with
glibc's getgrnam():
  disk, cdrom, floppy, tape, audio, video, lp, tty, dialout, kmem.
LDAP setups need to make sure, that these groups are always resolvable at
bootup, with only the rootfs mounted, and without network access available.

Some systems may need to add some new, currently not used groups, or need
to add some users to new groups, but the cost of this change is minimal,
compared to the pain the current, rather random, differences between the
various distributions cause for upstream projects and third-party vendors.

In general, "normal" users who log into a machine should never be a member
of any such group, but the device-access should be managed by dynamic ACLs,
which get added and removed for the specific users on login/logout and
session activity/inactivity. These groups are only provided for custom setups,
and mainly system services, to allow proper privilege separation.
A video-streaming daemon uid would be a member of "audio" and "video", to get
access to the sound and video devices, but no "normal" user should ever belong
to the "audio" group, because he could listen to the built-in microphone with
any ssh-session established from the other side of the world.

/dev/serial/by-{id,path}/ now contains links for ttyUSB devices,
which do not depend on the kernel device name. As usual, unique
devices - only a single one per product connected, or a real
USB serial number in the device - are always found with the same
name in the by-id/ directory.
Completely identical devices may overwrite their names in by-id/
and can only be found reliably in the by-path/ directory. Devices
specified by by-path/ must not change their connection, like the
USB port number they are plugged in, to keep their name.

To support some advanced features, Linux 2.6.22 is the oldest supported
version now. The kernel config with enabled SYSFS_DEPRECATED is no longer
supported. Older kernels should still work, and devices nodes should be
reliably created, but some rules and libudev will not work correctly because
the old kernels do not provide the expected information or interfaces.

udev 135
========
Bugfixes.

Fix for a possible segfault while swapping network interface names in udev
versions 131-134.

udev 134
========
Bugfixes.

The group "video" is part of the default rules now.

udev 133
========
Bugfix for kernels using SYSFS_DEPRECATED* option and finding parent
block devices in some cases. No common distro uses this option anymore,
and we do not get enough testing for this and recent udev versions. If
this option is not needed to run some old distro with a new kernel,
it should be disabled in the kernel config.

Bugfix for the $links substitution variable, which may crash if no links
are created. This should not happen in usual setups because we always
create /dev/{block,char}/ links.

The strings of the parsed rules, which are kept in memory, no longer
contain duplicate entries, or duplicate tails of strings. This, and the
new rules parsing/matching code reduces the total in-memory size of
a huge distro rule sets to 0.08 MB, compared to the 1.2MB of udev
version 130.

The export of DEVTYPE=disk/partition got removed from the default
rules. This value is available from the kernel. The pnp shell script
modprobe hack is removed from the default rules. ACPI devices have _proper_
modalias support and take care of the same functionality.
Installations which support old kernels, but install current default
udev rules may want to add that to the compat rules file.

Libvolume_id now always probes for all known filesystems, and does not
stop at the first match. Some filesystems are marked as "exclusive probe",
and if any other filesytem type matches at the same time, libvolume_id
will, by default, not return any probing result. This is intended to prevent
mis-detection with conflicting left-over signatures found from earlier
file system formats. That way, we no longer depend on the probe-order
in case of multiple competing signatures. In some setups the kernel allows
to mount a volume with just the old filesystem signature still in place.
This may damage the new filesystem and cause data-loss, just by mounting
it. Because volume_id can not decide which one the correct signature is,
the wrong signatures need to be removed manually from the volume, or the
volume needs to be reformatted, to enable filesystem detection and possible
auto-mounting.

udev 132
========
Fix segfault if compiled without optimization and dbg() does not get
compiled out and uses variables which are not available.

udev 131
========
Bugfixes. (And maybe new bugs. :))


extras/usb_id/usb_id.c
/*
 * usb_id - identify an USB device
 *
 * Copyright (c) 2005 SUSE Linux Products GmbH, Germany
 *
 * Author:
 *	Hannes Reinecke <hare@suse.de>
 *
 *	This program is free software; you can redistribute it and/or modify it
 *	under the terms of the GNU General Public License as published by the
 *	Free Software Foundation version 2 of the License.
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <unistd.h>
#include <string.h>
#include <ctype.h>
#include <fcntl.h>
#include <errno.h>
#include <getopt.h>

#include "../../udev/udev.h"

int debug;

static void log_fn(struct udev *udev, int priority,
		   const char *file, int line, const char *fn,
		   const char *format, va_list args)
{
	if (debug) {
		fprintf(stderr, "%s: ", fn != NULL ? fn : file);
		vfprintf(stderr, format, args);
	} else {
		vsyslog(priority, format, args);
	}
}

static char vendor_str[64];
static char model_str[64];
static char model_str_enc[256];
static char vendor_str_enc[256];
static char serial_str[UTIL_NAME_SIZE];
static char packed_if_str[UTIL_NAME_SIZE];
static char revision_str[64];
static char type_str[64];
static char instance_str[64];

static int use_usb_info;
static int use_num_info;

static void set_usb_iftype(char *to, int if_class_num, size_t len)
{
	char *type = "generic";

	switch (if_class_num) {
	case 1:
		type = "audio";
		break;
	case 2: /* CDC-Control */
		break;
	case 3:
		type = "hid";
		break;
	case 5: /* Physical */
		break;
	case 6:
		type = "media";
		break;
	case 7:
		type = "printer";
		break;
	case 8:
		type = "storage";
		break;
	case 9:
		type = "hub";
		break;
	case 0x0a: /* CDC-Data */
		break;
	case 0x0b: /* Chip/Smart Card */
		break;
	case 0x0d: /* Content Security */
		break;
	case 0x0e:
		type = "video";
		break;
	case 0xdc: /* Diagnostic Device */
		break;
	case 0xe0: /* Wireless Controller */
		break;
	case 0xfe: /* Application-specific */
		break;
	case 0xff: /* Vendor-specific */
		break;
	default:
		break;
	}
	strncpy(to, type, len);
	to[len-1] = '\0';
}

static int set_usb_mass_storage_ifsubtype(char *to, const char *from, size_t len)
{
	int type_num = 0;
	char *eptr;
	char *type = "generic";

	type_num = strtoul(from, &eptr, 0);
	if (eptr != from) {
		switch (type_num) {
		case 2:
			type = "atapi";
			break;
		case 3:
			type = "tape";
			break;
		case 4: /* UFI */
		case 5: /* SFF-8070i */
			type = "floppy";
			break;
		case 1: /* RBC devices */
			type = "rbc";
			break;
		case 6: /* Transparent SPC-2 devices */
			type = "scsi";
			break;
		default:
			break;
		}
	}
	util_strlcpy(to, type, len);
	return type_num;
}

static void set_scsi_type(char *to, const char *from, size_t len)
{
	int type_num;
	char *eptr;
	char *type = "generic";

	type_num = strtoul(from, &eptr, 0);
	if (eptr != from) {
		switch (type_num) {
		case 0:
		case 0xe:
			type = "disk";
			break;
		case 1:
			type = "tape";
			break;
		case 4:
		case 7:
		case 0xf:
			type = "optical";
			break;
		case 5:
			type = "cd";
			break;
		default:
			break;
		}
	}
	util_strlcpy(to, type, len);
}

#define USB_DT_DEVICE			0x01
#define USB_DT_INTERFACE		0x04

static int dev_if_packed_info(struct udev_device *dev, char *ifs_str, size_t len)
{
	char *filename = NULL;
	int fd;
	ssize_t size;
	unsigned char buf[18 + 65535];
	unsigned int pos, strpos;
	struct usb_interface_descriptor {
		u_int8_t	bLength;
		u_int8_t	bDescriptorType;
		u_int8_t	bInterfaceNumber;
		u_int8_t	bAlternateSetting;
		u_int8_t	bNumEndpoints;
		u_int8_t	bInterfaceClass;
		u_int8_t	bInterfaceSubClass;
		u_int8_t	bInterfaceProtocol;
		u_int8_t	iInterface;
	} __attribute__((packed));
	int err = 0;

	if (asprintf(&filename, "%s/descriptors", udev_device_get_syspath(dev)) < 0) {
		err = -1;
		goto out;
	}
	fd = open(filename, O_RDONLY);
	if (fd < 0) {
		fprintf(stderr, "error opening USB device 'descriptors' file\n");
		err = -1;
		goto out;
	}
	size = read(fd, buf, sizeof(buf));
	close(fd);
	if (size < 18 || size == sizeof(buf)) {
		err = -1;
		goto out;
	}

	pos = 0;
	strpos = 0;
	while (pos < sizeof(buf) && strpos+7 < len) {
		struct usb_interface_descriptor *desc;
		char if_str[8];

		desc = (struct usb_interface_descriptor *) &buf[pos];
		if (desc->bLength < 3)
			break;
		pos += desc->bLength;

		if (desc->bDescriptorType != USB_DT_INTERFACE)
			continue;

		if (snprintf(if_str, 8, "%02x%02x%02x:",
			     desc->bInterfaceClass,
			     desc->bInterfaceSubClass,
			     desc->bInterfaceProtocol) != 7)
			continue;

		if (strstr(ifs_str, if_str) != NULL)
			continue;

		memcpy(&ifs_str[strpos], if_str, 8),
		strpos += 7;
	}
out:
	free(filename);
	return err;
}

/*
 * A unique USB identification is generated like this:
 *
 * 1.) Get the USB device type from InterfaceClass and InterfaceSubClass
 * 2.) If the device type is 'Mass-Storage/SPC-2' or 'Mass-Storage/RBC'
 *     use the SCSI vendor and model as USB-Vendor and USB-model.
 * 3.) Otherwise use the USB manufacturer and product as
 *     USB-Vendor and USB-model. Any non-printable characters
 *     in those strings will be skipped; a slash '/' will be converted
 *     into a full stop '.'.
 * 4.) If that fails, too, we will use idVendor and idProduct
 *     as USB-Vendor and USB-model.
 * 5.) The USB identification is the USB-vendor and USB-model
 *     string concatenated with an underscore '_'.
 * 6.) If the device supplies a serial number, this number
 *     is concatenated with the identification with an underscore '_'.
 */
static int usb_id(struct udev_device *dev)
{
	struct udev *udev = udev_device_get_udev(dev);
	struct udev_device *dev_interface = NULL;
	struct udev_device *dev_usb = NULL;
	const char *if_class, *if_subclass;
	int if_class_num;
	int protocol = 0;

	dbg(udev, "syspath %s\n", udev_device_get_syspath(dev));

	/* shortcut, if we are called directly for a "usb_device" type */
	if (udev_device_get_devtype(dev) != NULL && strcmp(udev_device_get_devtype(dev), "usb_device") == 0) {
		dev_if_packed_info(dev, packed_if_str, sizeof(packed_if_str));
		dev_usb = dev;
		goto fallback;
	}

	/* usb interface directory */
	dev_interface = udev_device_get_parent_with_subsystem_devtype(dev, "usb", "usb_interface");
	if (dev_interface == NULL) {
		info(udev, "unable to access usb_interface device of '%s'\n",
		     udev_device_get_syspath(dev));
		return 1;
	}

	if_class = udev_device_get_sysattr_value(dev_interface, "bInterfaceClass");
	if (!if_class) {
		info(udev, "%s: cannot get bInterfaceClass attribute\n",
		     udev_device_get_sysname(dev));
		return 1;
	}

	if_class_num = strtoul(if_class, NULL, 16);
	if (if_class_num == 8) {
		/* mass storage */
		if_subclass = udev_device_get_sysattr_value(dev_interface, "bInterfaceSubClass");
		if (if_subclass != NULL)
			protocol = set_usb_mass_storage_ifsubtype(type_str, if_subclass, sizeof(type_str)-1);
	} else {
		set_usb_iftype(type_str, if_class_num, sizeof(type_str)-1);
	}

	info(udev, "%s: if_class %d protocol %d\n",
	     udev_device_get_syspath(dev_interface), if_class_num, protocol);

	/* usb device directory */
	dev_usb = udev_device_get_parent_with_subsystem_devtype(dev_interface, "usb", "usb_device");
	if (!dev_usb) {
		info(udev, "unable to find parent 'usb' device of '%s'\n",
		     udev_device_get_syspath(dev));
		return 1;
	}

	/* all interfaces of the device in a single string */
	dev_if_packed_info(dev_usb, packed_if_str, sizeof(packed_if_str));

	/* mass storage : SCSI or ATAPI */
	if ((protocol == 6 || protocol == 2) && !use_usb_info) {
		struct udev_device *dev_scsi;
		const char *scsi_model, *scsi_vendor, *scsi_type, *scsi_rev;
		int host, bus, target, lun;

		/* get scsi device */
		dev_scsi = udev_device_get_parent_with_subsystem_devtype(dev, "scsi", "scsi_device");
		if (dev_scsi == NULL) {
			info(udev, "unable to find parent 'scsi' device of '%s'\n",
			     udev_device_get_syspath(dev));
			goto fallback;
		}
		if (sscanf(udev_device_get_sysname(dev_scsi), "%d:%d:%d:%d", &host, &bus, &target, &lun) != 4) {
			info(udev, "invalid scsi device '%s'\n", udev_device_get_sysname(dev_scsi));
			goto fallback;
		}

		/* Generic SPC-2 device */
		scsi_vendor = udev_device_get_sysattr_value(dev_scsi, "vendor");
		if (!scsi_vendor) {
			info(udev, "%s: cannot get SCSI vendor attribute\n",
			     udev_device_get_sysname(dev_scsi));
			goto fallback;
		}
		udev_util_encode_string(scsi_vendor, vendor_str_enc, sizeof(vendor_str_enc));
		udev_util_replace_whitespace(scsi_vendor, vendor_str, sizeof(vendor_str)-1);
		udev_util_replace_chars(vendor_str, NULL);

		scsi_model = udev_device_get_sysattr_value(dev_scsi, "model");
		if (!scsi_model) {
			info(udev, "%s: cannot get SCSI model attribute\n",
			     udev_device_get_sysname(dev_scsi));
			goto fallback;
		}
		udev_util_encode_string(scsi_model, model_str_enc, sizeof(model_str_enc));
		udev_util_replace_whitespace(scsi_model, model_str, sizeof(model_str)-1);
		udev_util_replace_chars(model_str, NULL);

		scsi_type = udev_device_get_sysattr_value(dev_scsi, "type");
		if (!scsi_type) {
			info(udev, "%s: cannot get SCSI type attribute\n",
			     udev_device_get_sysname(dev_scsi));
			goto fallback;
		}
		set_scsi_type(type_str, scsi_type, sizeof(type_str)-1);

		scsi_rev = udev_device_get_sysattr_value(dev_scsi, "rev");
		if (!scsi_rev) {
			info(udev, "%s: cannot get SCSI revision attribute\n",
			     udev_device_get_sysname(dev_scsi));
			goto fallback;
		}
		udev_util_replace_whitespace(scsi_rev, revision_str, sizeof(revision_str)-1);
		udev_util_replace_chars(revision_str, NULL);

		/*
		 * some broken devices have the same identifiers
		 * for all luns, export the target:lun number
		 */
		sprintf(instance_str, "%d:%d", target, lun);
	}

fallback:
	/* fallback to USB vendor & device */
	if (vendor_str[0] == '\0') {
		const char *usb_vendor = NULL;

		if (!use_num_info)
			usb_vendor = udev_device_get_sysattr_value(dev_usb, "manufacturer");

		if (!usb_vendor)
			usb_vendor = udev_device_get_sysattr_value(dev_usb, "idVendor");

		if (!usb_vendor) {
			info(udev, "No USB vendor information available\n");
			return 1;
		}
		udev_util_encode_string(usb_vendor, vendor_str_enc, sizeof(vendor_str_enc));
		udev_util_replace_whitespace(usb_vendor, vendor_str, sizeof(vendor_str)-1);
		udev_util_replace_chars(vendor_str, NULL);
	}

	if (model_str[0] == '\0') {
		const char *usb_model = NULL;

		if (!use_num_info)
			usb_model = udev_device_get_sysattr_value(dev_usb, "product");

		if (!usb_model)
			usb_model = udev_device_get_sysattr_value(dev_usb, "idProduct");

		if (!usb_model) {
			dbg(udev, "No USB model information available\n");
			return 1;
		}
		udev_util_encode_string(usb_model, model_str_enc, sizeof(model_str_enc));
		udev_util_replace_whitespace(usb_model, model_str, sizeof(model_str)-1);
		udev_util_replace_chars(model_str, NULL);
	}

	if (revision_str[0] == '\0') {
		const char *usb_rev;

		usb_rev = udev_device_get_sysattr_value(dev_usb, "bcdDevice");
		if (usb_rev) {
			udev_util_replace_whitespace(usb_rev, revision_str, sizeof(revision_str)-1);
			udev_util_replace_chars(revision_str, NULL);
		}
	}

	if (serial_str[0] == '\0') {
		const char *usb_serial;

		usb_serial = udev_device_get_sysattr_value(dev_usb, "serial");
		if (usb_serial) {
			udev_util_replace_whitespace(usb_serial, serial_str, sizeof(serial_str)-1);
			udev_util_replace_chars(serial_str, NULL);
		}
	}
	return 0;
}

int main(int argc, char **argv)
{
	static const struct option options[] = {
		{ "usb-info", no_argument, NULL, 'u' },
		{ "num-info", no_argument, NULL, 'n' },
		{ "export", no_argument, NULL, 'x' },
		{ "debug", no_argument, NULL, 'd' },
		{ "help", no_argument, NULL, 'h' },
		{}
	};
	struct udev *udev;
	struct udev_device *dev = NULL;
	char syspath[UTIL_PATH_SIZE];
	const char *devpath;
	static int export;
	int retval = 0;

	udev = udev_new();
	if (udev == NULL)
		goto exit;

	logging_init("usb_id");
	udev_set_log_fn(udev, log_fn);

	while (1) {
		int option;

		option = getopt_long(argc, argv, "dnuxh", options, NULL);
		if (option == -1)
			break;

		switch (option) {
		case 'd':
			debug = 1;
			if (udev_get_log_priority(udev) < LOG_INFO)
				udev_set_log_priority(udev, LOG_INFO);
			break;
		case 'n':
			use_num_info = 1;
			use_usb_info = 1;
			break;
		case 'u':
			use_usb_info = 1;
			break;
		case 'x':
			export = 1;
			break;
		case 'h':
			printf("Usage: usb_id [--usb-info] [--num-info] [--export] [--help] <devpath>\n"
			       "  --usb-info  use usb strings instead\n"
			       "  --num-info  use numerical values\n"
			       "  --export    print values as environment keys\n"
			       "  --help      print this help text\n\n");
		default:
			retval = 1;
			goto exit;
		}
	}

	devpath = argv[optind];
	if (devpath == NULL) {
		fprintf(stderr, "No device specified\n");
		retval = 1;
		goto exit;
	}

	util_strlcpy(syspath, udev_get_sys_path(udev), sizeof(syspath));
	util_strlcat(syspath, devpath, sizeof(syspath));
	dev = udev_device_new_from_syspath(udev, syspath);
	if (dev == NULL) {
		err(udev, "unable to access '%s'\n", devpath);
		return 1;
	}

	retval = usb_id(dev);
	if (retval == 0) {
		char serial[256];

		util_strlcpy(serial, vendor_str, sizeof(serial));
		util_strlcat(serial, "_", sizeof(serial));
		util_strlcat(serial, model_str, sizeof(serial));
		if (serial_str[0] != '\0') {
			util_strlcat(serial, "_", sizeof(serial));
			util_strlcat(serial, serial_str, sizeof(serial));
		}
		if (instance_str[0] != '\0') {
			util_strlcat(serial, "-", sizeof(serial));
			util_strlcat(serial, instance_str, sizeof(serial));
		}

		if (export) {
			printf("ID_VENDOR=%s\n", vendor_str);
			printf("ID_VENDOR_ENC=%s\n", vendor_str_enc);
			printf("ID_MODEL=%s\n", model_str);
			printf("ID_MODEL_ENC=%s\n", model_str_enc);
			printf("ID_REVISION=%s\n", revision_str);
			printf("ID_SERIAL=%s\n", serial);
			if (serial_str[0] != '\0')
				printf("ID_SERIAL_SHORT=%s\n", serial_str);
			if (type_str[0] != '\0')
				printf("ID_TYPE=%s\n", type_str);
			if (instance_str[0] != '\0')
				printf("ID_INSTANCE=%s\n", instance_str);
			printf("ID_BUS=usb\n");
			if (packed_if_str[0] != '\0')
				printf("ID_USB_INTERFACES=:%s\n", packed_if_str);
		} else
			printf("%s\n", serial);
	}

exit:
	udev_device_unref(dev);
	udev_unref(udev);
	logging_close();
	return retval;
}
